RecoTracker/DeDx/plugins/DeDxHitCalibrator.cc

```diff
@@ -0,0 +1,392 @@
+                                       reco::DeDxHitCollection& stripHits) {
+  for (size_t i = 0; i < info.size(); i++) {
+    // Require hits to be complete and compatible
+    const auto& type = info.type(i);
+    if (!(type & (1 << reco::DeDxHitInfo::Complete)) || !(type & (1 << reco::DeDxHitInfo::Compatible)))
+      continue;
+
+    // Effective path length
+    const auto& pl = info.pathlength(i);
+    const auto pathLength = pl * (1. + 0.07 * std::log(pl / 450e-4));
+    const DetId& detId = info.detId(i);
+
+    // Strip
+    if (const auto& stripCluster = info.stripCluster(i)) {
+      const auto& thickness =
+          dynamic_cast<const StripGeomDetUnit*>(tkGeom_->idToDet(detId))->surface().bounds().thickness();
+      const auto& det = getDetId(detId, thickness);
+      const auto& chipId = ChipId(detId, stripCluster->barycenter() / 128.);
+      // Fill measured strip deposits
+      const auto& thr = dedxCalib_->thr[det];
+      std::vector<std::pair<double, int> > b;
+      b.reserve(stripCluster->amplitudes().size() + 2);
+      b.emplace_back(thr, isBelow);
+      for (const auto& adc : stripCluster->amplitudes()) {
+        if (adc > 253)
+          b.emplace_back(254., isOver);
+        else
+          b.emplace_back(adc + 0.5, isNormal);
+      }
+      b.emplace_back(thr, isBelow);
+      // Fit
+      const auto& result = fitStripCluster(b, dedxCalib_->alpha[det], 1. / dedxCalib_->sigma[det]);
+      // ADC -> e- -> MeV
+      const auto& energy = correctEnergy(result.first * 217 * 3.61e-6, chipId);
+      const auto& esigma = correctEnergy(result.second * 217 * 3.61e-6, chipId);
+      // Compute charge
+      const auto& charge = pathLength != 0. ? energy / pathLength : 0.;
+      stripHits.emplace_back(charge, trackMomentum, pathLength, 0, esigma);
+    }
+
+    // Pixel
+    if (const auto& pixelCluster = info.pixelCluster(i)) {
+      const auto& thickness =
+          dynamic_cast<const PixelGeomDetUnit*>(tkGeom_->idToDet(detId))->surface().bounds().thickness();
+      const auto& det = getDetId(detId, thickness);
+      const auto& chipId = ChipId(detId, (int(pixelCluster->x() / 80.) << 3) + int(pixelCluster->y() / 52.));
+      // Collect adc
+      double delta(0);
+      bool isSaturated(false);
+      for (size_t j = 0; j < pixelCluster->pixelADC().size(); j++) {
+        const auto& elec = pixelCluster->pixelADC()[j];
+        delta += elec * 3.61e-6;
+        // ( pos , (x , y) )
+        const auto& row = pixelCluster->minPixelRow() + pixelCluster->pixelOffset()[2 * j];
+        const auto& col = pixelCluster->minPixelCol() + pixelCluster->pixelOffset()[2 * j + 1];
+        // Go back to adc
+        const auto& DBgain = pixelCalib_.getGain(detId, col, row);
+        const auto& DBpedestal = pixelCalib_.getPedestal(detId, col, row);
+        if (DBgain > 0.) {
+          const auto vcal = (elec + 414.) / 65.;
+          const auto adc = DBpedestal + vcal / DBgain;
+          if (adc > 185)
```

Comment:
@sikler : What is the 185?

Reply:
The condition "adc > 185" was there to identify whether a channel was in overflow.
It is somewhat experimental, and seemed to be a good choice back that time.

I think there could be better ways, I wonder if pixel/strip experts have a more solid
definition for being in overflow.

Reply:
@sikler if you can't find those numbers coded somewhere else in the release, either make them configurable, or at least use a named `constexpr` to keep record of them.

Reply:
Maybe @ferencek, @mroguljic or @mmusich would know how to better determine if a pixel channel is in overflow

Reply:
I am not sure to see how adc > 185 can signal an overflow (in the pixels). ADC range is 8bits so the value can go from 0 to 255 (2^8-1). I would expected overflows in the last bit. What I am missing?

Reply:
I see that in https://github.com/cms-sw/cmssw/blob/master/CalibTracker/SiStripCommon/plugins/ShallowGainCalibration.cc#L185 the criteria of amplitude >= 254 is used to determine the condition of saturation for a pixel cluster

Reply:
> the criteria of amplitude >= 254 

I wouldn't trust that (experimental) code that is a straight copy of the section above (for Strips). In the strip detector, bits 254 and 255 have very precise meanings:

https://github.com/cms-sw/cmssw/blob/ff1f93a873b5f03ef2b8cb075954c7a75bcd75ba/SimTracker/SiStripDigitizer/src/SiTrivialDigitalConverter.cc#L82-L83

concerning different levels of saturation. In the pixels things are handled differently.

Reply:
Ok I see. So then, would it be better to simply check if adc == 255 when checking for overflow for pixels?

Reply:
> Ok I see. So then, would it be better to simply check if adc == 255 when checking for overflow for pixels?

Let me tag also @dkotlins

Reply:
Danek answered [here](https://mattermost.web.cern.ch/cms-exp/pl/1qtkaxbhi3fwipiytiwd8z61py) and for convenience I am quoting it here: "The upper references are for strips. The pixelADC() returns electrons.  Almost everything in the clusterizer is in electrons. ADC values (from digis) are translated to electrons in the early stage of the clusterization process in copy_to_buffer() method of the PixelThresholdClusterizer. To get ADC values one has to look at digis or apply the reversed gain calibration specific to each pixel."

Reply:
Ok, so if I understand correctly, the current approach of converting the electron charge to ADC (by inverting the vcal and pixel calibrations as currently done in the code) is the way to derive the ADC, which then should be used to derive the overflow of pixel hits

Reply:
Maybe it would be best to have 185 as a parameter(?)

I remember playing with that, since the back then the overflow did not correspond to ADC == 255, but to a smaller number. Could be related to the incorrent backward conversion from electrons to ADC though.

Reply:
Set 185 as a parameter in 75444097ae14d9911caf215c6bb9e407f0551754